### PR TITLE
QR Code Module

### PR DESCRIPTION
**Pull request by IceOnly**
With BC19.1 codeunit 4113 "Swiss QR Code Helper" is introduced.
This Codeunit use the mexican .net Object to create a qr code and overlay the swiss cross to the qr code.
It is not possible to call the pure QR code generation.

So I created this system module.

### CODE REVIEW

**Comment by JesperSchulz** [comment]
Thank you for your contribution! We will definitely take a look :-)

**Comment by Groenbech96** [comment]
I think this is good! 
Some thoughts I have:
This implementation only allows to go from text to an image. 
What if you wanted to go the other way? 

Also this implementation could use the Image Module for storing the image. 
What about a failure scenario for test? Are there any?

**Comment by IceOnly** [comment]
> I think this is good! Some thoughts I have: This implementation only allows to go from text to an image. What if you wanted to go the other way?

QR to Text is not implemented in the Microsoft.Dynamics.Nav.MX.dll assembly.

> Also this implementation could use the Image Module for storing the image.

Then what would be the advantages? Cropping or Resizing an QR code is not the best idea.

> What about a failure scenario for test? Are there any?

I added one.

**Comment by JesperSchulz** [comment]
First of all, great work on creating a new module! From a design perspective, this one is close to spot on :-)

We had some internal discussions though, as to which extend this code should be part of the Barcode module. Sure, the barcode module currently uses fonts to generate 1D and 2D barcodes, but as a developer, I would/should turn to _that_ module if I want to create a barcode of any sort - QR codes included. Ideally, we should not have competing modules which serve the same purpose. Initially, I recall that the barcode module was designed to support different kinds of barcode generation: fonts and images. I don't recall, if we removed that implementation for simplification purposes - I will have to look into that again. We'll get back to you, once we'd had a chance to look a little closer into, how these two modules could fit together!

**Comment by JesperSchulz** [comment]
So we just did look a little closer into this PR and concluded, that we should merge your code into the barcode module. It would only make sense to add the capability to generate QR codes there, which already is a symbology in the barcode module. Now we just generate them as images, rather as fonts.

On top of that, it should be fairly easy to merge your code into the barcode module. Let me try to explain how we envision that:
In the src folder, we currently have 4 folders; 2 of which are for 2D barcodes, such as this is. As this is a new provider, we should create 2 more folders: Barcode **Image** Provider 2D and Dynamics Barcode Provider (the name is up for discussion). In the BarcodeFontEncoder folder, you will find 4 files. The symbology file, we should move one level up - it is not specific to the font encoder. The other 3 files we should copy to the Image Provider folder and adjust for our barcode image implementation; now we return BLOBs, rather than text. In the new provider folder, we again replicate; this time the IDAutomation 2D provider. Here, the actual implementation goes, as already created in this PR. In the IDAutomation Encoders folder, we have support for 5 symbologies, for the Dynamics provider there will only be 1: QR Code. And that would be it! Now the barcode module also supports generation of QR code images - in the same way, as we generate barcode encoded text strings.

Confused? I think it's relatively simple, once you understand the design of the barcode module. All of your code and tests remain largely unchanged - it just needs to get moved around and adjusted slightly :-)

I hope you're up for the challenge to update your PR. If you'd rather not, I will offer you to do it for you. But it would of course be so much better if you would reel this PR in and fully keep it _your_ contribution!

And if you need help or even a short call, I'd be more than happy to further assist!

Looking forward to your feedback.

**Review by JesperSchulz — CHANGES_REQUESTED** [review]
As described in the previous post, we should merge this code into the barcode module. We should not have two competing modules, which essentially offer the same functionality. Details are in my previous post.

**Comment by IceOnly** [comment]
@JesperSchulz i created the first protoype to clarify the structure. I keep the QR Code Module. I will remove it after clearing the rest, also the new test are not implmented yet.

I would rather like to create two new folders:
"Barcode Provider 2D" and "Barcode Provider" and move "Barcode Font Provider" to "Barcode Provider", "Barcode Image Provider 2D" and "Barcode Font Provider 2D" to "Barcode Provider 2D".

And there is an additional problem I have. In my orginal Implementation it is possible to declare the correction level and version of the qr code. Any idea how to pass this information to the new implementation?

**Comment by IceOnly** [comment]
I added a parameter table "Barcode Encode Settings 2D" to control the QR Code specific values, changed the folder structure and added the tests.

Is there a recommendation  for atool to create the README.md from the XML documentation?

File: .gitignore [hunk 1]
```diff
@@ -294,3 +294,4 @@ AddOns/AMCBanking365Fundamentals/test/app.json
 AddOns/AMCBanking365Fundamentals/test/AppResources/Mock request  response/App.config
 AddOns/AMCBanking365Fundamentals/app/app.json
 AddOns/AMCBanking365Fundamentals/app/app.json
+**/.netpackages/**
```
Comment (JesperSchulz) [inline .gitignore:297]: Unrelated change. Please revert.

File: Modules/System/Barcode/src/Barcode Provider 2D/BarcodeEncodeSettings2D.Table.al [hunk 2]
```diff
@@ -0,0 +1,54 @@
+// ------------------------------------------------------------------------------------------------
+// Copyright (c) Microsoft Corporation. All rights reserved. 
+// Licensed under the MIT License. See License.txt in the project root for license information.
+// ------------------------------------------------------------------------------------------------
+
+/// <summary>
+/// Common setting used when encoding @D barcodes.
```
Comment (JesperSchulz) [inline Modules/System/Barcode/src/Barcode Provider 2D/BarcodeEncodeSettings2D.Table.al:7]: 2D rather than @D? And are these settings indeed for 2D barcodes, or are these specific to the QR code symbology?

File: Modules/System/Barcode/src/Barcode Provider 2D/Image/BarcodeImageEncoder2D.Interface.al [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+/// </summary>
+interface "Barcode Image Encoder 2D"
+{
+    /// <summary> 
+    /// Encodes a input text to a barcode image.
+    /// </summary>
+    /// <param name="InputText">The text to encode.</param>
+    /// <param name="BarcodeEncodeSettings2D">The settings to use when encoding the text.</param>
+    /// <returns>The encoded barcode.</returns>
+    procedure EncodeImage(InputText: Text; BarcodeEncodeSettings2D: Record "Barcode Encode Settings 2D"): Codeunit "Temp Blob";
```
Comment (JesperSchulz) [inline Modules/System/Barcode/src/Barcode Provider 2D/Image/BarcodeImageEncoder2D.Interface.al:17]: Would it not make sense to have an overload without the settings parameter table? Does one always have to explicitly define the options, or are there default settings, which often would do the trick?

File: Modules/System/Barcode/src/Barcode Provider 2D/BarcodeEncodeSettings2D.Table.al [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+/// </summary>
+table 9204 "Barcode Encode Settings 2D"
```
Comment (JesperSchulz) [inline Modules/System/Barcode/src/Barcode Provider 2D/BarcodeEncodeSettings2D.Table.al:9]: Are these settings indeed for 2D barcodes, or are they for your QR code implementation? I they are for QR code only, I'm not sure if I'm a fan...

File: Modules/System/Barcode/src/Barcode Provider 2D/Image/BarcodeImageProvider2D.enum.al [hunk 5]
```diff
@@ -0,0 +1,22 @@
+// ------------------------------------------------------------------------------------------------
```
Comment (JesperSchulz) [inline Modules/System/Barcode/src/Barcode Provider 2D/Image/BarcodeImageProvider2D.enum.al:1]: I don't think we should add another set of symbologies for the image provider. Reuse the existing one. Your implementation only supports the QR symbology, but that is alright.

**Review by JesperSchulz — CHANGES_REQUESTED** [review]
Nice work!
Left a couple of questions in the code review which we should go through, before I approve this PR. But it's definitely getting there!

File: Modules/System/Barcode/src/Barcode Provider 2D/BarcodeEncodeSettings2D.Table.al (hunk 4 above)
Comment (JesperSchulz) [reply Modules/System/Barcode/src/Barcode Provider 2D/BarcodeEncodeSettings2D.Table.al:9]: ...but then again, we did the same with our 1D barcodes as well. And I didn't particularly liked it there either ;-) But we should probably be consistent and do it here as well. So I guess I'm ok with this.

File: Modules/System/Barcode/src/Barcode Provider 2D/Image/BarcodeImageEncoder2D.Interface.al (hunk 3 above)
Comment (JesperSchulz) [reply Modules/System/Barcode/src/Barcode Provider 2D/Image/BarcodeImageEncoder2D.Interface.al:17]: I guess you copied the implementation we've got in 1D barcodes? In that case, we're probably ok.

File: .gitignore (hunk 1 above)
Comment (IceOnly) [reply .gitignore]: Done. Can I open another related PR for this? all 728 .net assemblies are shown as changes...

File: Modules/System/Barcode/src/Barcode Provider 2D/BarcodeEncodeSettings2D.Table.al (hunk 4 above)
Comment (IceOnly) [reply Modules/System/Barcode/src/Barcode Provider 2D/BarcodeEncodeSettings2D.Table.al:9]: I hate this solution. But as you mentioned it is the same way as the 1D implementation.

File: Modules/System/Barcode/src/Barcode Provider 2D/Image/BarcodeImageProvider2D.enum.al (hunk 5 above)
Comment (IceOnly) [reply Modules/System/Barcode/src/Barcode Provider 2D/Image/BarcodeImageProvider2D.enum.al:1]: I used the same symboligies. Or should I merge the providers?
enum 9206 "Barcode Font Provider 2D" and enum 9207 "Barcode Image Provider 2D" into one with two Implementations?
But this will be destructive if I rename the file to "Barcode Provider 2D".

File: Modules/System/Barcode/src/Barcode Provider 2D/BarcodeEncodeSettings2D.Table.al (hunk 2 above)
Comment (IceOnly) [reply Modules/System/Barcode/src/Barcode Provider 2D/BarcodeEncodeSettings2D.Table.al]: Typo fixed.

File: Modules/System/Barcode/src/Barcode Provider 2D/Image/BarcodeImageEncoder2D.Interface.al (hunk 3 above)
Comment (IceOnly) [reply Modules/System/Barcode/src/Barcode Provider 2D/Image/BarcodeImageEncoder2D.Interface.al:17]: There is already an overload in interface "Barcode Image Provider 2D".
Normally I wouldn't call the interface "Barcode Image Encoder 2D" directly, , would I?

File: .gitignore (hunk 1 above)
Comment (JesperSchulz) [reply .gitignore]: Sure you can :-) But we prefer to separate issues from each other. Makes PR management (e.g. backporting to earlier releases) SO much easier :-)

File: Modules/System/Barcode/src/Barcode Provider 2D/Image/BarcodeImageEncoder2D.Interface.al (hunk 3 above)
Comment (JesperSchulz) [reply Modules/System/Barcode/src/Barcode Provider 2D/Image/BarcodeImageEncoder2D.Interface.al:17]: Nope, we're good here.

File: Modules/System/Barcode/src/Barcode Provider 2D/BarcodeEncodeSettings2D.Table.al (hunk 4 above)
Comment (JesperSchulz) [reply Modules/System/Barcode/src/Barcode Provider 2D/BarcodeEncodeSettings2D.Table.al:9]: Then we're in agreement. We close our eyes and push forward ;-)

File: Modules/System/Barcode/src/Barcode Provider 2D/Image/BarcodeImageProvider2D.enum.al (hunk 5 above)
Comment (JesperSchulz) [reply Modules/System/Barcode/src/Barcode Provider 2D/Image/BarcodeImageProvider2D.enum.al:1]: Sorry, that was a "brain fart". I'm still struggling a little to get the overview in this code review tool. My bad. All good here.

**Comment by JesperSchulz** [comment]
Thanks for answering my questions so quickly. Just needed to double check that we're doing the right things here - we are! From my perspective, we're good to go! Let's get a 3rd pair of eyes on this PR and in it goes :-)

**Review by JesperSchulz — APPROVED** [review]

File: Modules/System Tests/Barcode/src/Dynamics2D/Dynamics2DQrCodeTest.Codeunit.al [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+    var
+        BarcodeImageProvider2D: Enum "Barcode Image Provider 2D";
+        IBarcodeImageProvider2D: Interface "Barcode Image Provider 2D";
+    begin
+        // [SCENARIO] A string variable is converted to qr code
+        IBarcodeImageProvider2D := BarcodeImageProvider2D::Dynamics2D;
+
+        // [WHEN] The string is converted
+        // [THEN] no error occured
+        IBarcodeImageProvider2D.EncodeImage(SampleTxt, Enum::"Barcode Symbology 2D"::"QR-Code");
```
Comment (mazhelez) [inline Modules/System Tests/Barcode/src/Dynamics2D/Dynamics2DQrCodeTest.Codeunit.al:27]: Another way to test that no error occurred is to clear the last error before the invocation of `EncodeImage` and then assert it is empty after the invocation.

File: Modules/System Tests/Barcode/src/Dynamics2D/Dynamics2DQrCodeTest.Codeunit.al [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+    begin
+        // [SCENARIO] A 1274 byte string is converted to an Version 40 QR code with High Error Correction Level (Max Capacity 1273 bytes)
+        IBarcodeImageProvider2D := BarcodeImageProvider2D::Dynamics2D;
+        BarcodeEncodeSettings2D.Init();
+        BarcodeEncodeSettings2D."Error Correction Level" := BarcodeEncodeSettings2D."Error Correction Level"::High;
+        BarcodeEncodeSettings2D."Module Size" := 1;
+
+        // [WHEN] The string is converted
+        // [THEN] error occured
+        asserterror IBarcodeImageProvider2D.EncodeImage(GetMaxAlphaNumericString(true), Enum::"Barcode Symbology 2D"::"QR-Code", BarcodeEncodeSettings2D);
```
Comment (mazhelez) [inline Modules/System Tests/Barcode/src/Dynamics2D/Dynamics2DQrCodeTest.Codeunit.al:64]: Missing `Assert.ExpectedError(...)`

File: Modules/System Tests/Barcode/src/Dynamics2D/Dynamics2DQrCodeTest.Codeunit.al [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+    end;
+
+    local procedure GetMaxAlphaNumericString(PlusOne: Boolean) Result: Text
+    var
+        I: Integer;
+    begin
+        for i := 1 to 1273 do
+            Result += 'A';
```
Comment (mazhelez) [inline Modules/System Tests/Barcode/src/Dynamics2D/Dynamics2DQrCodeTest.Codeunit.al:72]: Does it have to be "AA...A"?

Otherwise, consider using codeunit `Any` from the test lib `Any`.

**Review by mazhelez — APPROVED** [review]
Nicely done!

A few touches on the tests and the PR will be perfect :)

File: Modules/System Tests/Barcode/src/Dynamics2D/Dynamics2DQrCodeTest.Codeunit.al (hunk 8 above)
Comment (IceOnly) [reply Modules/System Tests/Barcode/src/Dynamics2D/Dynamics2DQrCodeTest.Codeunit.al]: Done,

File: Modules/System Tests/Barcode/app.json [hunk 9]
```diff
@@ -28,6 +28,12 @@
                              "name":  "Barcode",
                              "publisher":  "Microsoft",
                              "version":  "19.2.0.0"
+                         },
+                         {
+                             "id": "e31ad830-3d46-472e-afeb-1d3d35247943",
+                              "name": "BLOB Storage",
```
[the remark this answers is not in the document]
Comment (IceOnly) [reply Modules/System Tests/Barcode/app.json:40]: Now it is used. :-)

File: Modules/System Tests/Barcode/src/Dynamics2D/Dynamics2DQrCodeTest.Codeunit.al (hunk 7 above)
Comment (IceOnly) [reply Modules/System Tests/Barcode/src/Dynamics2D/Dynamics2DQrCodeTest.Codeunit.al:70]: Done.

File: Modules/System Tests/Barcode/src/Dynamics2D/Dynamics2DQrCodeTest.Codeunit.al (hunk 6 above)
Comment (IceOnly) [reply Modules/System Tests/Barcode/src/Dynamics2D/Dynamics2DQrCodeTest.Codeunit.al]: Done

**Comment by JesperSchulz** [comment]
Looks like we're done here then :-) I will now process this PR in our internal build system. If all goes well, this can get backported to our 19.x release branch tomorrow and then we're ready for release with 19.3! Great job!

**Comment by achim-t** [comment]
I tried to use this.

```al
    local procedure CreateBarcodeImage() BarcodeQR: Text
    var
        BarcodeImageProvider2D: Interface "Barcode Image Provider 2D";
        TempBlob: Codeunit "Temp Blob";
        is: InStream;
    begin
        BarcodeImageProvider2D := Enum::"Barcode Image Provider 2D"::Dynamics2D;
        TempBlob := BarcodeImageProvider2D.EncodeImage('hello world!', "Barcode Symbology 2D"::"QR-Code");
        TempBlob.CreateInStream(is);
        is.Read(BarcodeQR);
    end;
```

The content of BarcodeQr is `'BMZ�'`, which does not seem to be an image. 
What did I do wrong?

**Comment by IceOnly** [comment]
> I tried to use this.
> 
> ```al
>     local procedure CreateBarcodeImage() BarcodeQR: Text
>     var
>         BarcodeImageProvider2D: Interface "Barcode Image Provider 2D";
>         TempBlob: Codeunit "Temp Blob";
>         is: InStream;
>     begin
>         BarcodeImageProvider2D := Enum::"Barcode Image Provider 2D"::Dynamics2D;
>         TempBlob := BarcodeImageProvider2D.EncodeImage('hello world!', "Barcode Symbology 2D"::"QR-Code");
>         TempBlob.CreateInStream(is);
>         is.Read(BarcodeQR);
>     end;
> ```
> 
> The content of BarcodeQr is `'BMZ�'`, which does not seem to be an image. What did I do wrong?

@achim-t i don't understand that you are trying. 
You read binary data in your text variable. Read will stop at the first zero byte. It is starting with BM so it is a valid bitmap.

try this code and you will see it works fine:

```al
   local procedure CreateBarcodeImage()
    var
        BarcodeImageProvider2D: Interface "Barcode Image Provider 2D";
        FileManagement: Codeunit "File Management";
        TempBlob: Codeunit "Temp Blob";
    begin
        BarcodeImageProvider2D := Enum::"Barcode Image Provider 2D"::Dynamics2D;
        TempBlob := BarcodeImageProvider2D.EncodeImage('hello world!', "Barcode Symbology 2D"::"QR-Code");
        FileManagement.BLOBExport(TempBlob, 'qrcode.bmp', true);
    end;
```

**Comment by achim-t** [comment]
@IceOnly  
Thanks for verifying that the bitmap is indeed in the TempBlob.
I want to print the QR on a report, so I was trying to get it out of the TempBlob. 

It works using Base64Convert:

```al
local procedure CreateBarcodeImage(txt: Text) Result: Text
var
    TempBlob: Codeunit "Temp Blob";
    Base64Convert: Codeunit "Base64 Convert";
    BarcodeImageProvider2D: Interface "Barcode Image Provider 2D";
    is: InStream;
begin
    BarcodeImageProvider2D := Enum::"Barcode Image Provider 2D"::Dynamics2D;
    TempBlob := BarcodeImageProvider2D.EncodeImage(txt, "Barcode Symbology 2D"::"QR-Code");
    TempBlob.CreateInStream(is);
    Result := Base64Convert.ToBase64(is);
end;
```
